src/main/java/ArrayIndexedBag.java

```diff
@@ -0,0 +1,174 @@
+        for (int i = 0; i < bag.length; ++i) {
+            newBag[i] = bag[i];
+        }
+        bag = newBag;
+    }
+
+    /**
+     * Shifts elements in an array down (towards index 0) towards
+     * the starting index specified. This method assumes that
+     * the calling function manages the rear field.
+     *
+     * @param start Index of where shift starts/stops and element
+     *              overwritten.
+     */
+    private void shiftLeft(int start) {
+        for (int i = start; i < bag.length - 1; ++i) {
+            bag[i] = bag[i + 1];
+        }
+        bag[bag.length - 1] = null;
+    }
+
+    /**
+     * Shifts elements in an array up (towards index rear) away
+     * from the starting index specified. This method assumes that
+     * the calling function manages the rear field and any
+     * need of expandCapacity().
+     *
+     * @param start Index of where shifting starts/stops and where
+     *              space is opened.
+     */
+    private void shiftRight(int start) {
+        for (int i = rear; i > start; --i) {
+            bag[i] = bag[i - 1];
+        }
+        bag[start] = null;
+    }
+
+    @Override
+    public void add(T element) {
+        add(rear, element);
+    }
+
+    @Override
+    public void add(int index, T element) {
+        if (index > rear) {
+            throw new IndexOutOfBoundsException(String.format("Bag has no index %d to add to.", index));
+        }
+        if (size() == bag.length) {
+            expandCapacity();
+        }
+        shiftRight(index);
+        bag[index] = element;
+        rear++;
+    }
+
+    @Override
+    public void set(int index, T element) {
+        if (index >= rear) {
+            throw new IndexOutOfBoundsException(String.format("Bag has no element at index %d.", index));
+        }
+        bag[index] = element;
+    }
+
+    @Override
+    public T get(int index) {
+        if (index >= rear) {
+            throw new IndexOutOfBoundsException(String.format("Bag has no element at index %d.", index));
+        }
+        return bag[index];
+    }
+
+    @Override
+    public T remove(int index) {
+        if (index >= rear) {
+            throw new IndexOutOfBoundsException(String.format("Bag has no element at index %d.", index));
+        }
+        T returnElement = bag[index];
+        shiftLeft(index);
+        rear--;
+        return returnElement;
+    }
+
+    @Override
+    public int indexOf(T target) {
+        for (int i = 0; i < bag.length; ++i) {
+            if (bag[i].equals(target)) {
+                return i;
+            }
+        }
+        return NOT_FOUND;
+    }
+
+    @Override
+    public T remove(T element) {
+        if (isEmpty()) {
+            throw new NoSuchElementException("Removing from an empty bag.");
+        }
+        int removeIndex = indexOf(element);
+        if (removeIndex == NOT_FOUND) {
+            throw new NoSuchElementException("Element not contained in bag.");
+        }
```

Comment:
You have conflicting method signatures with this. `indexOf` returns a special `NOT_FOUND` value, whereas `remove` throws when something is not found. Stick with the principal of least surprise, if you want sentinels, use sentinels. If you want exceptions, use exceptions.

Reply:
:facepalm: Javadoc I wrote even says throw exception.

Reply:
Well. Reading is hard, you know that.